docs/configuration.md

```diff
@@ -278,13 +279,22 @@ domain must be separated by a comma.
       mapping:
         printer.lan: 192.168.178.3
         otherdevice.lan: 192.168.178.15,2001:0db8:85a3:08d3:1319:8a2e:0370:7344
-        anothername.lan: CNAME(otherdevice.lan)
+      zone: |
+        $ORIGIN example.com.
+        www 3600 A 1.2.3.4
+        @ 3600 CNAME www
     ```
 
 This configuration will also resolve any subdomain of the defined domain, recursively. For example querying any of
 `printer.lan`, `my.printer.lan` or `i.love.my.printer.lan` will return 192.168.178.3.
 
-CNAME records are supported by setting the value of the mapping to `CNAME(target)`. Note that the target will be recursively resolved and will return an error if a loop is detected.
+CNAME records are supported by utilizing the `zone` parameter. The zone file is a multiline string containing a [DNS Zone File](https://en.wikipedia.org/wiki/Zone_file#Example_file).
+For records defined using the `zone` parameter, the `customTTL` parameter is unused. Instead, the TTL is defined in the zone directly.
+The following directives are supported in the zone file:
+* `$ORIGIN` - sets the origin for relative domain names
+* `$TTL` - sets the default TTL for records in the zone
+* `$INCLUDE` - includes another zone file relative to the blocky executable
```

Comment:
I think it would be better UX to make it relative to the config path, my bad for not realizing this before.

It requires a bit of a hack because the Go YAML unmarshaling API doesn't let you easily pass in additional data (also applies to pretty much any other Go unmarshaling API).  

The idea is to make `ZoneFileDNS` a struct with two fields:
```go
type ZoneFileDNS struct{
	RRs []dns.RR

	configPath string // only used during unmarshaling
}
```

And then in `loadConfig`, we can set `configPath` before calling `unmarshalConfig`, so that `ZoneFileDNS.UnmarshalYAML` can read `z.configPath`.

Here's what that code should look like (additions highlighted):
```go
var (
	data       []byte
	prettyPath = path // addition
)

if fs.IsDir() {
	prettyPath = filepath.Join(path, "*") // addition

	data, err = readFromDir(path, data)
	if err != nil {
		return nil, fmt.Errorf("can't read config files: %w", err)
	}
} else {
	data, err = os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("can't read config file: %w", err)
	}
}

cfg.CustomDNS.Zone.configPath = prettyPath // addition

err = unmarshalConfig(logger, data, &cfg)
```

And finally in `UnmarshalYAML` we can do `dns.NewZoneParser(strings.NewReader(input), "", z.configPath)`.

The downside of this technique is you need a struct with a field that's only used during unmarshaling, but I think it's worth it in this case.

Let me know if I can clarify anything because it's a bit of a long winded comment.

Reply:
This was super helpful. Thank you! I ended up writing 2 additional tests to cover bare filenames (Files in the same dir), relative filenames, `./other.zone`, and kept the original absolute path test so that we can ensure however a user defines their zone file, we _should_ have reasonable assurances that it will correctly resolve.